### PR TITLE
Ziemleszcz/ipsec

### PR DESCRIPTION
**Pull request by ziemleszcz**
## Description
IPSEC headers and getprotobyname/getprotobynumber implementation.

## Motivation and Context
Required for IPSEC driver and openiked port.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (PC).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: sys/proto.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+
+struct protoent_internal {
+	const char *name;
+	const char *aliases[MAX_ALIASES];
+	int proto;
+};
+
+
+struct protoent_internal protocols[PROTOCOLS_NUM] = {
```
Comment (agkaminski) [inline sys/proto.c:32]: Could be static const

File: sys/proto.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * %LICENSE%
+ */
+
+#include <netdb.h>
+#include <string.h>
+#include <stdlib.h>
+
+
+#define PROTOCOLS_NUM 55
```
Comment (agkaminski) [inline sys/proto.c:21]: It's better and safer to get this number by using `sizeof(protocols) / sizeof(*protocols)`

File: sys/proto.c [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+static void protoent_free(struct protoent *proto)
+{
+	unsigned int i = 0;
+
+	if (proto->p_name)
+		free(proto->p_name);
+
+	if (proto->p_aliases) {
+		while (i < MAX_ALIASES && proto->p_aliases[i]) {
```
Comment (agkaminski) [inline sys/proto.c:103]: This while works exactly once - for `i == 0`. Is this intended? What about `p_aliases[1]` and `p_aliases[2]` (`p_aliases` have 3 indexes)?

Comment (agkaminski) [reply sys/proto.c:103]: Also, `for` loop fits here better

File: sys/proto.c [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+static void protoent_copy(struct protoent *proto, unsigned int num)
+{
+	unsigned int i = 0;
+
+	proto->p_name = strdup(protocols[num].name);
+
+	while (i < MAX_ALIASES && protocols[num].aliases[i]) {
+		proto->p_aliases[i] = strdup(protocols[num].aliases[i]);
+		++i;
+	}
```
Comment (agkaminski) [inline sys/proto.c:120]: As above, only one iteration, for loop fits better

File: sys/proto.c [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+	proto->p_proto = protocols[num].proto;
+}
+
+
+struct protoent *getprotobyname(const char *name)
+{
+	unsigned int i = 0;
+
+	while (i < PROTOCOLS_NUM) {
```
Comment (agkaminski) [inline sys/proto.c:131]: For loop fits better

File: sys/proto.c [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+	return NULL;
+}
+
+
+struct protoent *getprotobynumber(int proto)
+{
+	unsigned int i = 0;
+
+	while (i < PROTOCOLS_NUM) {
```
Comment (agkaminski) [inline sys/proto.c:148]: For loop fits better

**Review by agkaminski — CHANGES_REQUESTED** [review]

File: sys/proto.c (hunk 3 above)
Comment (ziemleszcz) [reply sys/proto.c]: It loops 0 to MAX_ALIASES times. `p_aliases` is array of string pointers terminated by NULL. This function is used only on `proto_common` variable.

**Review by agkaminski — APPROVED** [review]
Searching perhaps could be done better (right now is O(n), could be O(1) by number and O(log n) by name), but if we don't need peak performance LGTM.

**Pull request merged**